Implement SourceUnitEnumChunker for GitLab

mcastorina opened the pull request:
<!--
Please create an issue to collect feedback prior to feature additions. Please also reference that issue in any PRs.
If possible try to keep PRs scoped to one feature, and add tests for new features.
-->

### Description:
Implements `SourceUnitEnumChunker` in GitLab for enumerating and chunking units.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

mcastorina committed:
Implement SourceUnitEnumChunker for GitLab

mcastorina commented on pkg/sources/gitlab/gitlab.go:
```diff
@@ -406,8 +428,17 @@ func (s *Source) getAllProjects(ctx context.Context, apiClient *gitlab.Client) (
+	ignoreRepo func(string) bool,
+	reporter sources.UnitReporter,
+) ([]string, error) {
+	projects, err := s.getAllProjects(ctx, apiClient, reporter)
```
It would be nice if this could stream the projects so we can start getting the repos early, but I'll leave that for a future optimization.

mcastorina committed:
Add GitLab engine integration test

rosecodym commented on pkg/engine/gitlab_integration_test.go:
```diff
@@ -0,0 +1,41 @@
+	// Check the output provided by metrics.
+	metrics := e.GetMetrics()
+	assert.GreaterOrEqual(t, metrics.ChunksScanned, uint64(36390))
+	assert.GreaterOrEqual(t, metrics.BytesScanned, uint64(92662889))
```
Are you using `GreaterOrEqual` to guard against future stuff being added to Gitlab?

rosecodym commented on pkg/sources/gitlab/gitlab.go:
```diff
@@ -344,7 +355,11 @@ func (s *Source) getAllProjects(ctx context.Context, apiClient *gitlab.Client) (
+			if err := reportErr(ctx, err); err != nil {
+				return nil, err
+			}
+			break
```
Does using `break` instead of `return` mean that this error is now non-fatal when running with units?

mcastorina replied:
Yes, it's non-fatal for both units and non-units. I thought it made sense to do so instead of quitting early.

rosecodym replied:
Ah right, both cases, derp. I think it's fine to do, just wanted to ensure I understood.

rosecodym commented on pkg/sources/gitlab/gitlab.go:
```diff
@@ -368,7 +383,11 @@ func (s *Source) getAllProjects(ctx context.Context, apiClient *gitlab.Client) (
+			if err := reportErr(ctx, err); err != nil {
+				return nil, err
+			}
+			break
```
same question about `break`

mcastorina replied:
Yep, this was intentional. Do you think it's okay to do this? I'll double check the flow.

rosecodym commented on pkg/sources/gitlab/gitlab.go:
```diff
@@ -564,3 +606,77 @@ func normalizeRepos(repos []string) ([]string, []error) {
+	// Otherwise, enumerate all repos.
+	ignoreRepo := buildIgnorer(s.ignoreRepos, func(err error, pattern string) {
+		ctx.Logger().Error(err, "could not compile ignore repo glob", "glob", pattern)
+		// TODO: Handle error returned from UnitErr.
```
Any reason you can't just return it? An un-compilable ignore glob sounds pretty fatal to me.

mcastorina replied:
It's in a callback right now, so it's just adding the error to the report. This comment is referring to the error when adding the error to the report, which can happen from context cancellation. If that's the case, it'll throw the same error when returning later in this function, so I don't think it's worth the effort to escape the error(s) out of the closure.

But you bring up a good point about un-compilable globs - should we make it fatal? Right now it appears they are not.

rosecodym replied:
Oh, right, derp again. It's probably a good thing to handle at some point but I don't think you need to do it as part of this PR. (I believe it's written that way so that `Validate` can detect it, so it's not like it's completely undiscoverable.)

rosecodym reviewed (commented):
this mostly looks good but i have a small question about some control flow

rosecodym reviewed (approved):
Looks pretty straightforward. I see that you're accommodating the legacy `Chunks` workflow by passing a `nil` reporter and then checking for reporter nil-ness. One idea I'm playing around with in the Confluence sources is creating a local reporter that does the "legacy" logging, rather than just passing `nil`. Is that something you considered?

mcastorina commented:
> Looks pretty straightforward. I see that you're accommodating the legacy `Chunks` workflow by passing a `nil` reporter and then checking for reporter nil-ness. One idea I'm playing around with in the Confluence sources is creating a local reporter that does the "legacy" logging, rather than just passing `nil`. Is that something you considered?

Ooh interesting. So calling `ChunksOk` on the reporter would send the chunk on the channel, and `ChunksErr` would log the error? I like it.

That reminded me of the [ChanReporter](https://github.com/trufflesecurity/trufflehog/blob/v3.67.3/pkg/sources/chan_reporter.go#L10-L13), and it turns out it already does exactly that lol. I should use it.

What about for `UnitOk` and `UnitErr`? `UnitOk` appends to a slice maybe? 🤔

mcastorina committed:
Use a SliceReporter instead of checking for nil reporters

mcastorina committed:
Use more generic VisitorReporter

mcastorina committed:
Merge logic from getReposFromGitlab into getAllProjectRepos

mcastorina committed:
Update integration test to have a lower bound

Unfortunately, the GitLab integration test does not appear to be
deterministic. Sometimes 36390 chunks are found, sometimes 36312, or
even lower.

rosecodym commented on pkg/sources/gitlab/gitlab.go:
```diff
@@ -246,19 +251,24 @@ func (s *Source) Validate(ctx context.Context) []error {
-		if !ignoreProject(p.PathWithNamespace) {
-			return errs
-		}
+	if len(repos) == 0 {
+		errs = append(errs, fmt.Errorf("ignore patterns excluded all projects"))
```
lol this is way clearer than my old code that you changed, thanks!

rosecodym commented on pkg/sources/gitlab/gitlab.go:
```diff
@@ -310,28 +320,54 @@ func (s *Source) basicAuthSuccessful(apiClient *gitlab.Client) bool {
-				uniqueProjects[proj.ID] = proj
-				projects = append(projects, proj)
-				projectsWithNamespace = append(projectsWithNamespace, proj.NameWithNamespace)
+			// Skip projects we've already seen.
```
Any idea what causes this?

mcastorina replied:
Not sure what you're referring to here. If you're referring to seeing projects we've already seen, then I _think_ it's because we're paginating so there might be some overlap in rare cases.

rosecodym commented on pkg/sources/gitlab/gitlab.go:
```diff
@@ -564,3 +587,74 @@ func normalizeRepos(repos []string) ([]string, []error) {
+}
+
+// ChunkUnit downloads and reports chunks for the given GitLab repository unit.
+func (s *Source) ChunkUnit(ctx context.Context, unit sources.SourceUnit, reporter sources.ChunkReporter) error {
```
Did you consider trying to unify this code path with the one in `scanRepos`? I ask because it includes some non-trivial logic that seems easy to forget about during later maintenance. (But I can see the argument against it: that some error handling and logic makes unifying them non-trivial itself.)

mcastorina replied:
There was a bit too much logic around progress / concurrency / error handling that it was easier to copy/paste and delete.

I'll try to unify in a followup PR.

rosecodym reviewed (approved):
lgtm!
